<a href="https://colab.research.google.com/github/gerega37/mipt_study/blob/main/Olesya_HW3_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Разогрев

## Задание 1

Иногда, когда нужно создавать много однотипных лямбда функций, еще удобнее будет создать функцию, которая будет их генерировать.

Реализуйте функцию ``mod_checker(x, mod=0)``, которая будет **возвращать лямбда функцию** от одного аргумента ``y``, которая будет возвращать ``True``, если остаток от деления ``y`` на ``x`` равен ``mod``, и ``False`` иначе.

Если в ``return`` вашей функции написано выражение, отличное от лямбда функции - задание считается проваленным

**Sample:**

```python
mod_3 = mod_checker(3)

print(mod_3(3)) # True
print(mod_3(4)) # False

mod_3_1 = mod_checker(3, 1)
print(mod_3_1(4)) # True
```

In [ ]:
def mod_checker(x, mod=0):
    """ Return True if the remainder of dividing y by x is mod. """
    return lambda y: y % x == mod

In [ ]:
assert mod_checker(3)(3) == True
assert mod_checker(3)(4) == False
assert mod_checker(3, 1)(4) == True

# Numpy

In [ ]:
import numpy as np

## Задание 2
Напишите функцию, которая принимает один аргумент - ``n``. Функция должна создать нампаевскую матрицу размером ``(n, n)`` и залить ее шахматным узором. Порядок узора не имеет значения, важен общий паттерн.

**Sample:**
```python
chess_maze(8)

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]
```



In [ ]:
def chess_maze(n: int) -> np.ndarray:
    """
    Create an array with a checkerboard pattern.

    param: n is a number of elements in a row
    return: an array with a checkerboard pattern
    """
    zeros = np.zeros((n,n), dtype=np.int8)
    zeros[1:n:2, 0:n:2] = 1
    zeros[0:n:2, 1:n:2] = 1
    return zeros


print(chess_maze(8))

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]


## Задание 3
Напишите функцию, которая принимает 2 аргумента ``arr`` и ``exclude_values``. ``arr`` - это одномерный нампаевский массив, в котором необходимо поменять знак всем числам (инвертировать), которые есть в массиве ``exclude_values``. Действия необходимо делать прямо в исходном массиве, не создавая копий.

**Sample:**
```python
#Исходный вектор 
arr = np.arange(10+1)
exclude_values = np.arange(4, 8)
 
my_func(arr, exclude_values)
array([ 0  1  2  3 -4 -5 -6 -7  8  9 10])
```

In [ ]:
def my_func(arr: np.ndarray, exclude_values: np.ndarray) -> np.ndarray:
    """
    Change a sign of elements in an array which are included in exclude_values.

    param: arr is an input array,
    exclude_value is an array with elements that need to be changed
    return: arr is a modified array
    """
    for i in range(len(arr)):
        if arr[i] in exclude_values:
            arr[i] = 0 - arr[i]
    return arr

In [ ]:
assert (my_func(np.arange(2, 8), np.arange(4, 6)) == np.array([2, 3, -4, -5, 6, 7])).all()

In [ ]:
assert [1, 2, 3] == [1, 2 ,3]

In [ ]:
assert (np.array([1, 2, 3]) == np.array([1, 2 ,3])).all()

## Задание 4
Напишите функцию, которая принимает одномерный положительный нампаевский массив и минимум 4 раза выводит на экран (функция ``print``) этот же массив, в котором у элементов оставлена только целая часть. Все >= 4 раза функция должна использовать разные методы для извлечения целой части.

**Sample:**
```python
#Исходный вектор 
arr = np.random.uniform(low=10, high=100,size=20)
#Исходный вектор arr:
arr
[99.54707368 49.26237526 80.50565073 10.10998238 16.39378916 60.64633657
 25.49637622 31.06574268 94.48636127 73.41204506 75.70638137 70.43830146
 72.28607422 40.59319738 57.85802554 35.39428061 55.25707514 48.00020145
 38.39449991 26.19085017]
 
my_func(arr)

[99. 49. 80. 10. 16. 60. 25. 31. 94. 73. 75. 70. 72. 40. 57. 35. 55. 48.
 38. 26.]
[99. 49. 80. 10. 16. 60. 25. 31. 94. 73. 75. 70. 72. 40. 57. 35. 55. 48.
 38. 26.]
[99. 49. 80. 10. 16. 60. 25. 31. 94. 73. 75. 70. 72. 40. 57. 35. 55. 48.
 38. 26.]
[99 49 80 10 16 60 25 31 94 73 75 70 72 40 57 35 55 48 38 26]
[99. 49. 80. 10. 16. 60. 25. 31. 94. 73. 75. 70. 72. 40. 57. 35. 55. 48.
 38. 26.]
```

В примере 5 выводов print из функции, просто чтобы были :)

In [ ]:
def integer(arr: np.ndarray) -> np.ndarray:
    """ Get rid of the fractional part of the floating point number. """
    arr1 = arr.astype(dtype=np.uint16)
    arr2 = np.trunc(arr)
    arr3 = arr.astype(dtype=np.int16)
    arr4 = np.fix(arr)
    return arr, arr1, arr2, arr3, arr4


arr = np.random.uniform(low=10, high=100,size=20)

(array([12.00726767, 80.06962269, 16.75581765, 97.15236324, 95.87410625,
        74.9615048 , 63.51032519, 94.89847967, 62.97277243, 57.14334473,
        84.84933379, 67.71494054, 30.66491535, 72.00892834, 78.8844412 ,
        50.23323795, 36.07963429, 96.51711838, 21.34879403, 74.2341196 ]),
 array([12, 80, 16, 97, 95, 74, 63, 94, 62, 57, 84, 67, 30, 72, 78, 50, 36,
        96, 21, 74], dtype=uint16),
 array([12., 80., 16., 97., 95., 74., 63., 94., 62., 57., 84., 67., 30.,
        72., 78., 50., 36., 96., 21., 74.]),
 array([12, 80, 16, 97, 95, 74, 63, 94, 62, 57, 84, 67, 30, 72, 78, 50, 36,
        96, 21, 74], dtype=int16),
 array([12., 80., 16., 97., 95., 74., 63., 94., 62., 57., 84., 67., 30.,
        72., 78., 50., 36., 96., 21., 74.]))

## Задание 5
Напишите функцию, которая принимает два числа - ``N`` и ``p``. ``N`` - это размер стороны квадратной матрицы np.zeros, ``p`` - это количество индексов в матрице (p не превышает N*N), которые нужно заменить на случайные числа. Индексы и числа нужно выбирать случайно. Функция должна возвращать квадратную матрицу со стороной ``N``, в которой ровно ``p`` чисел заменены на случайные.

```python
my_func(N=6, p=4)

[[56.   0.   0.   0.   0.   0.]
 [  0.   0.   0.   0.   0.   0.]
 [99.   0.   0.   0.   0.   0.]
 [  0.   0.   0. 83.   0.   0.]
 [  0.   0.   0.   0.   0.   0.]
 [  0.   0.   0.   0.   1.   0.]]
```

In [ ]:
def my_func(N: int, p: int) -> np.ndarray:
    """
    Return an array with values changed to random numbers.

    :param: N is the side of the square array,
    p is a number of numbers to be replaced by random ones
    :return: arr is an array with changed values
    """
    arr = np.zeros((N, N))
    for i in range(p):
        number = np.random.randint(100)
        index1 = np.random.randint(N)
        index2 = np.random.randint(N)
        arr[index1][index2] = number
    return arr


my_func(6,4)

# Pandas

In [ ]:
import pandas as pd

# Задание 6

Любым удобным для вас способом создайте ``dataframe`` c именем ``my_data``, в котором две колонки c именами (``type`` - строки, ``value`` - числа с плавающей точкой) и четыре наблюдения в каждой колонке:

```
type value
B    18.

B    20.

C    10.

C    -10.
```

Какой размер созданного вами датафрейма? (Подумайте, что лучше использовать - информацию из метода df.info() или информацию из метода df.memory_usage()?)
1) Измените тип данных колонки type на category. Как изменился размер датафрейма?
2) Измените тип данных колонки value на менее жадный. Как изменился размер датафрейма?

In [ ]:
my_data = pd.DataFrame(
    data=[
      ['B', 18],
      ['B', 20],
      ['C', 10],
      ['C', -10]
    ],
    columns=['type', 'value'],
    dtype=(str, float)
)

In [ ]:
my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   type    4 non-null      object
 1   value   4 non-null      object
dtypes: object(2)
memory usage: 192.0+ bytes


In [ ]:
my_data.memory_usage()

Index    128
type      32
value     32
dtype: int64

In [ ]:
my_data['type'] = my_data['type'].astype('category')
my_data.info()
my_data.memory_usage()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   type    4 non-null      category
 1   value   4 non-null      object  
dtypes: category(1), object(1)
memory usage: 288.0+ bytes


Index    128
type     128
value     32
dtype: int64

In [ ]:
my_data['value'] = my_data['value'].astype('float16')
my_data.info()
my_data.memory_usage()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   type    4 non-null      category
 1   value   4 non-null      float16 
dtypes: category(1), float16(1)
memory usage: 264.0 bytes


Index    128
type     128
value      8
dtype: int64

The size of the original data frame is 192.0+ bytes. After changing the data type of the column 'type' to 'category', the size of the data frame increased to 288.0+ bytes. Changing the data type of the column 'data' to float16 reduced the data frame to 264.0 bytes.

# Задание 7
Любым удобным для вас способом создайте ``dataframe`` c именем ``my_data``, в котором две колонки c именами (``type`` - строки, ``value`` - числа с плавающей точкой) и десять тысяч (10 000) наблюдений в каждой колонке. В колонке ``type`` должно быть 10 уникальных значений.

```python
my_data.shape
____________
(10000, 2)
```

Какой размер созданного вами датафрейма? (Подумайте, что лучше использовать - информацию из метода df.info() или информацию из метода df.memory_usage()?)
1) Измените тип данных колонки type на category. Как изменился размер датафрейма?
2) Измените тип данных колонки value на менее жадный. Как изменился размер датафрейма?

Во сколько раз уменьшился размер исходного датафрейма после выполнения этих двух операций?

In [ ]:
value_data = np.arange(10000, dtype=float)
type_data = map(str, np.random.randint(0, 10, 10000))
d = {'type': type_data, 'value': value_data}
my_data = pd.DataFrame(data=d)
display(my_data)

type   value
0       3     0.0
1       4     1.0
2       8     2.0
3       4     3.0
4       5     4.0
...   ...     ...
9995    5  9995.0
9996    4  9996.0
9997    1  9997.0
9998    0  9998.0
9999    7  9999.0

[10000 rows x 2 columns]

In [ ]:
my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   type    10000 non-null  object 
 1   value   10000 non-null  float64
dtypes: float64(1), object(1)
memory usage: 156.4+ KB


In [ ]:
# df.memory_usage() helps us to understand which column takes up more memory
my_data.memory_usage()

Index      128
type     80000
value    80000
dtype: int64

In [ ]:
my_data['type'] = my_data['type'].astype('category')
my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   type    10000 non-null  category
 1   value   10000 non-null  float64 
dtypes: category(1), float64(1)
memory usage: 88.4 KB


In [ ]:
my_data['value'] = my_data['value'].astype('float16')
# the original data frame has decreased by more than 5 times
my_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   type    10000 non-null  category
 1   value   10000 non-null  float16 
dtypes: category(1), float16(1)
memory usage: 29.8 KB


The size of the original data frame is 156.4+ KB. After changing the data type of the column 'type' to 'category', the size of the data frame decreased to 88.4+ KB. Changing the data type of the column 'data' to float16 reduced the data frame to 29.8 KB.



#### Для решения дальнейших задач мы будем использовать следующий датасет, загрузите его с помощью команды:

In [ ]:
df = pd.read_csv('https://stepik.org/media/attachments/course/4852/my_stat.csv')

In [ ]:
df

Примечание: все операции с исходным датафреймом должны быть ``inplace``. То есть исходный датафрейм в переменной ``df`` **должен** измениться

# Задание 8

В переменную с именем subset_1 сохраните только первые 19 строк и только 1 и 3 колонку.

В переменную с именем subset_2 сохраните все строки кроме 2, 3 и 18 и только 2 и 4 колонку.

In [ ]:
subset_1 = df.iloc[:19, [0,2]]

In [ ]:
subset_2 = df.iloc[:, [1,3]]
subset_2 = subset_2.drop([1,2,17])

# Задание 9

В переменную subset_3 сохраните только те наблюдения, у которых значения переменной V1 строго больше 0, и значение переменной V3  равняется 'A'.

В переменную  subset_4  сохраните только те наблюдения, у которых значения переменной V2  не равняются 10, или значения переменной V4 больше или равно 1.

In [ ]:
mask1 = subset_1['V1'] > 0
mask2 = subset_1['V3'] == 'A'
subset_3 = subset_1[mask1 & mask2]

In [ ]:
mask3 = subset_2['V2'] != 10
mask4 = subset_2['V4'] >= 1
subset_4 = subset_2[mask3 | mask4]

# Задание 10
Добавьте следующие колонки в df:

1) V5 = V1 + V4
2) V6 = **натуральный** логарифм переменной V2

In [ ]:
df['V5'] = df['V1'] + df['V4']
df['V6'] =  np.log(df['V2'])

# Задание 11

Переименуйте колонки в df.

1) V1 -> session_value

2) V2 -> group

3) V3 -> time

4) V4 -> n_users

In [ ]:
df.rename(
    {"V1": "session_value", "V2": "group", "V3": "time", "V4": "n_users"},
          axis="columns")

# Задание 12
В датафрейме ``my_stat`` в колонке ``session_value`` замените все пропущенные значения на нули.

В колонке ``n_users`` замените все отрицательные значения на медианное значение переменной ``n_users`` (без учета отрицательных значений, разумеется).

In [ ]:
url = 'https://stepik.org/media/attachments/course/4852/my_stat_1.csv'
my_stat = pd.read_csv(url)

In [ ]:
# replace all missing values with zero
my_stat = my_stat.fillna(0)

In [ ]:
negative_values = my_stat['n_users'] < 0
my_stat[negative_values] = my_stat[my_stat['n_users'] > 0]['n_users'].median()